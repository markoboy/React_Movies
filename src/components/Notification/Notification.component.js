import { faCheckCircle, faTimesCircle } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import React, { useMemo } from 'react';

export default function NotificationComponent({
  title,
  message,
  isError,
}) {
  const icon = useMemo(() => (isError ? faTimesCircle : faCheckCircle), [isError]);

  return (
    <div className="text-align--center">
      <FontAwesomeIcon className="color--primary" size="3x" icon={icon} />
      <h1 className="font--regular text-transform--uppercase">{title}</h1>
      <p>{message}</p>
    </div>
  );
}

NotificationComponent.defaultProps = {
  message: '',
  isError: false,
};

NotificationComponent.propTypes = {
  title: PropTypes.string.isRequired,
  message: PropTypes.string,
  isError: PropTypes.bool,
};
