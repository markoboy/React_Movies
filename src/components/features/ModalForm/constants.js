export const MODAL_ADD_TITLE = 'Add Movie';
export const MODAL_EDIT_TITLE = 'Edit Movie';
export const MODAL_DELETE_TITLE = 'Delete Movie';

export const MODAL_FORM_ADD_ACTION = 'Add';
export const MODAL_FORM_EDIT_ACTION = 'Edit';
export const MODAL_FORM_DELETE_ACTION = 'Delete';

export const MODAL_FORM_DELETE_PARAGRAPH = 'Are you sure you want to delete this movie?';

export const GENRE_SELECT_OPTIONS = [
  {
    label: 'Documentary',
    value: 'documentary',
  },
  {
    label: 'Comedy',
    value: 'comedy',
  },
  {
    label: 'Horror',
    value: 'horror',
  },
  {
    label: 'Crime',
    value: 'crime',
  },
];
