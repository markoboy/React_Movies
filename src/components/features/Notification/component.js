import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import React from 'react';

export default function NotificationComponent({ title, message, icon }) {
  return (
    <div className="text-align--center">
      <FontAwesomeIcon className="color--primary" size="3x" icon={icon} />
      <h1 className="font--regular text-transform--uppercase">{title}</h1>
      <p>{message}</p>
    </div>
  );
}

NotificationComponent.defaultProps = {
  message: '',
};

NotificationComponent.propTypes = {
  title: PropTypes.string.isRequired,
  message: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
  icon: PropTypes.object.isRequired,
};
