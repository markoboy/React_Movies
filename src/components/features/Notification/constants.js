export const SUCCESS_TITLE = 'Congratulations!';
export const ERROR_TITLE = 'Oops an Error occurred!';
