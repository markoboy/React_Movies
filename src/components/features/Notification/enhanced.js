import {
  moviesErrorSelector,
  moviesStatusSelector,
  moviesSuccessSelector,
} from '@store/selectors/moviesSelectors';
import { connect } from 'react-redux';
import NotificationContainer from './container';

const mapStateToProps = (state) => ({
  error: moviesErrorSelector(state),
  success: moviesSuccessSelector(state),
  status: moviesStatusSelector(state),
});

export default connect(mapStateToProps)(NotificationContainer);
