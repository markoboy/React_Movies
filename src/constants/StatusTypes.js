const StatusTypes = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  COMPLETE: 'COMPLETE',
  ERROR: 'ERROR',
};

export default StatusTypes;
