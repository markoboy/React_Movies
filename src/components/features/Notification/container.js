/* eslint-disable react/forbid-prop-types */
import Modal from '@components/common/Modal';
import { COMPLETE_STATUS, ERROR_STATUS } from '@constants/StatusTypes';
import { faCheckCircle } from '@fortawesome/free-solid-svg-icons/faCheckCircle';
import { faTimesCircle } from '@fortawesome/free-solid-svg-icons/faTimesCircle';
import useNotification from '@hooks/UseNotification';
import PropTypes from 'prop-types';
import React, { useCallback, useEffect, useMemo } from 'react';
import NotificationComponent from './component';
import { ERROR_TITLE, SUCCESS_TITLE } from './constants';

export default function NotificationContainer({ error, success, status, removeNotification }) {
  const {
    isOpened,
    setIsOpened,
    notification,
    setNotification,
  } = useNotification();

  useEffect(() => {
    if (status === ERROR_STATUS) {
      setNotification(error, ERROR_TITLE);
    } else if (status === COMPLETE_STATUS) {
      setNotification(success, SUCCESS_TITLE);
    }
  }, [success, error, status]);

  const icon = useMemo(
    () => (status === ERROR_STATUS ? faTimesCircle : faCheckCircle),
    [isOpened]
  );

  const closeTriggerHandler = useCallback(() => {
    setIsOpened(false);
    removeNotification();
  }, []);

  return isOpened ? (
    <Modal
      onCloseTrigger={closeTriggerHandler}
      body={<NotificationComponent icon={icon} {...notification} />}
    />
  ) : null;
}

NotificationContainer.defaultProps = {
  error: null,
  success: null,
};

NotificationContainer.propTypes = {
  error: PropTypes.object,
  success: PropTypes.object,
  status: PropTypes.string.isRequired,
  removeNotification: PropTypes.func.isRequired,
};
