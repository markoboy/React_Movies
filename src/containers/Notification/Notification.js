import Modal from '@components/Modal';
import { ERROR_STATUS } from '@constants/StatusTypes';
import { moviesErrorSelector, moviesStatusSelector, moviesSuccessSelector } from '@store/selectors/moviesSelectors';
import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import ErrorNotification from './ErrorNotification';
import SuccessNotification from './SuccessNotification';

function Notification({
  error,
  success,
  status,
}) {
  const [isOpened, setIsOpened] = useState(false);

  const [modalBody, setModalBody] = useState(null);

  useEffect(() => {
    setModalBody(
      status === ERROR_STATUS
        ? <ErrorNotification message={error?.message} title={error?.title} />
        : <SuccessNotification message={success?.message} title={success?.title} />
    );

    setIsOpened(!!success || !!error);
  }, [success, error, status]);

  return (
    <>
      {isOpened && (
        <Modal
          onCloseTrigger={() => setIsOpened(false)}
          body={modalBody}
        />
      )}
    </>
  );
}

const mapStateToProps = (state) => ({
  error: moviesErrorSelector(state),
  success: moviesSuccessSelector(state),
  status: moviesStatusSelector(state),
});

export default connect(mapStateToProps)(Notification);
