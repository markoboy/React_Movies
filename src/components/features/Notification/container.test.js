import React from 'react';
import { render } from '@testing-library/react';

import { COMPLETE_STATUS, ERROR_STATUS } from '@constants/StatusTypes';
import { ERROR_TITLE, SUCCESS_TITLE } from './constants';
import Notification from './container';

describe('Notification Component', () => {
  let removeNotification;

  beforeEach(() => {
    removeNotification = jest.fn();
  });

  it('renders success message when status is COMPLETE_STATUS', () => {
    const success = {
      message: 'Successful message',
    };
    const { container } = render(
      <Notification
        removeNotification={removeNotification}
        success={success}
        status={COMPLETE_STATUS}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders error message when status is ERROR_STATUS', () => {
    const error = {
      message: 'Error message',
    };
    const { container } = render(
      <Notification
        removeNotification={removeNotification}
        error={error}
        status={ERROR_STATUS}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
