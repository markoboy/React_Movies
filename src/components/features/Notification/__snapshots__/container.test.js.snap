// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Component renders error message when status is ERROR_STATUS 1`] = `
<div>
  <div
    class="modal"
    data-testid="modal"
  >
    <div
      class="modal__content"
    >
      <button
        class="modal__close-trigger"
        data-testid="modal-close-trigger"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-times fa-w-11 fa-2x "
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="modal__body"
      >
        <div
          class="text-align--center"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-times-circle fa-w-16 fa-3x color--primary"
            data-icon="times-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
              fill="currentColor"
            />
          </svg>
          <h1
            class="font--regular text-transform--uppercase"
          >
            Oops an Error occurred!
          </h1>
          <p>
            Error message
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Notification Component renders success message when status is COMPLETE_STATUS 1`] = `
<div>
  <div
    class="modal"
    data-testid="modal"
  >
    <div
      class="modal__content"
    >
      <button
        class="modal__close-trigger"
        data-testid="modal-close-trigger"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-times fa-w-11 fa-2x "
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="modal__body"
      >
        <div
          class="text-align--center"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check-circle fa-w-16 fa-3x color--primary"
            data-icon="check-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              fill="currentColor"
            />
          </svg>
          <h1
            class="font--regular text-transform--uppercase"
          >
            Congratulations!
          </h1>
          <p>
            Successful message
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
